pxr/usd/usd/stageCache.cpp

```diff
@@ -65,69 +58,153 @@ std::atomic_long idCounter(9223000);
 Id GetNextId() { return Id::FromLongInt(++idCounter); }
 
 struct Entry {
-    Entry() {}
+    Entry() = default;
     Entry(const UsdStageRefPtr &stage, Id id) : stage(stage), id(id) {}
     UsdStageRefPtr stage;
     Id id;
 };
 
-struct ById {};
+using StagesById = std::unordered_map<Id, UsdStageRefPtr, TfHash>;
+using IdsByStage = std::unordered_map<UsdStageRefPtr, Id, TfHash>;
+using StagesByRootLayer = std::unordered_multimap<SdfLayerHandle,
+                                                  UsdStageRefPtr,
+                                                  TfHash>;
+
+/// The stage container provides a set of synchronization maps to enable
+/// lookup by Id, LayerHandle, or Stage. It provides a single insertion
+/// function to populate these maps and erasure by multiple key types.
+///
+/// This data structure does not add any thread safety guarantees. Users should
+/// lock appropriately before read and write.
+class StageContainer {
+public:
+    StageContainer() = default;
+
+    /// Return an unordered_map of the stages indexed by id
+    const StagesById& ById() const { return _byId; }
+    /// Return an unordered_map of the ids indexed by stage
+    const IdsByStage& ByStage() const { return _byStage; }
+    /// Return an unordered_multimap of the stages indexed by root layer
+    /// NOTE: This is a multimap where multiple stages may mapped by a single
+    /// root layer.
+    const StagesByRootLayer& ByRootLayer() const { return _byRootLayer; }
+
+    // If the stage is already in the cache, return the pair {Id, false}
+    // If the stage is not in the cache, generate a new id, populate
+    // the synchronized maps, and return {NewId, true}
+    std::pair<Id, bool> Insert(const UsdStageRefPtr& stage) {
+        const auto it = _byStage.find(stage);
+        if (  it != _byStage.cend()) {
+            return std::make_pair(it->second, false);
+        }
+        const Id id = GetNextId();
+        TF_AXIOM(_byStage.emplace(stage, id).second);
+        TF_AXIOM(_byId.emplace(id, stage).second);
+        _byRootLayer.emplace(stage->GetRootLayer(), stage);
+        return std::make_pair(id, true);
+    }
 
-struct ByStage {};
+    // Return true if the stage was successfully erased
+    bool Erase(const UsdStageRefPtr& stage) {
+        const auto it = _byStage.find(stage);
```

Comment:
It's preferred that the `return false;` is on a separate line and wrapped in curly braces.